[WGSL] Add type declarations for logical built-in functions

tadeuzagallo opened the pull request:
#### 8c713be450f29a1b8e0e7c78a09b756c5cdd7cea
<pre>
[WGSL] Add type declarations for logical built-in functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=255513">https://bugs.webkit.org/show_bug.cgi?id=255513</a>
rdar://108134231

Reviewed by Mike Wyrzykowski.

Add type declarations for the 3 functions from section 17.3 (Logical Built-in Functions) of the spec[1].

[1]: <a href="https://www.w3.org/TR/WGSL/#logical-builtin-functions">https://www.w3.org/TR/WGSL/#logical-builtin-functions</a>

* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: <a href="https://commits.webkit.org/263064@main">https://commits.webkit.org/263064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f7d16853cf08d0729d063ef7136d58d9a066bca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3360 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4785 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3691 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3486 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3444 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2920 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3401 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3684 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3029 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4604 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1192 "1 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2992 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2921 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2965 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3055 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4344 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3432 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2759 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2995 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2993 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/848 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3000 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WGSL/TypeDeclarations.rb:
```diff
@@ -164,3 +164,20 @@
+
+# 17.3.1 & 17.3.2
+["all", "any"].each do |op|
+    operator :"#{op}", {
```
Q: "all" and "any" are strings and then we interpolate them in `:"#{op}". Is there a way to avoid the string interpolation? I couldn't make it work when I tried but my knowledge of Ruby is only at an intermediate level.

tadeuzagallo replied:
My first thought is that we can tweak `operator` to accept either a string or a symbol (it might even work today, but not sure). A quick google also suggests that we can use `op.to_sym`. I'll land this as is, and refactor all the use cases later.

djg reviewed (approved):
:shipit:

tadeuzagallo committed:
[WGSL] Add type declarations for logical built-in functions
https://bugs.webkit.org/show_bug.cgi?id=255513
rdar://108134231

Reviewed by Mike Wyrzykowski.

Add type declarations for the 3 functions from section 17.3 (Logical Built-in Functions) of the spec[1].

[1]: https://www.w3.org/TR/WGSL/#logical-builtin-functions

* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: https://commits.webkit.org/263064@main
